apps/web/src/core/providers/web-vm.ts

```diff
@@ -0,0 +1,388 @@
+    import("xterm"),
+    import("xterm-addon-webgl"),
+    import("xterm-addon-fit")
+  ])
+
+  const terminal = new Terminal({
+    convertEol: true,
+    cursorBlink: true,
+    allowProposedApi: true
+  })
+
+  const fitAddon = new FitAddon()
+  terminal.loadAddon(new WebglAddon())
+  terminal.loadAddon(fitAddon)
+  terminal.open(el)
+  fitAddon.fit()
+
+  return {
+    terminal,
+    fitAddon
+  }
+}
+
+const WORKING_BIN_PATH = ".bin"
+const JSH_HISTORY_PATH = "/home/.jsh_history"
+
+// Use this bin cache for temp memory cache
+// TODO: can probably read the file itself OR make this into a map to store the AI-generated script itself
+const binCache = new Set<string>()
+
+const addBin = async (cmd: string, container: WebContainer) => {
+  // Use this to read past cmd history if needed:
+  // const proc = await container.spawn("cat", [JSH_HISTORY_PATH])
+  // const { value: historyEntries } = await proc.output.getReader().read()
+
+  if (binCache.has(cmd)) {
+    return
+  }
+
+  binCache.add(cmd)
+  // Just make empty file for now
+  const cmdBinPath = posix.join(WORKING_BIN_PATH, cmd)
+  await container.fs.writeFile(cmdBinPath, "")
+  const chmodProc = await container.spawn("chmod", ["+x", cmdBinPath], {
+    output: false
+  })
+
+  await chmodProc.exit
+}
+
+// TODO: Make this a cool map or something
+const customCommandList = ["ai", "code"] as const
+type CustomCommand = (typeof customCommandList)[number]
+
+const customCommandSet = new Set(customCommandList)
+const isCustomCommand = (cmd: any): cmd is CustomCommand =>
+  customCommandSet.has(cmd)
+
+// You will think step-by-step for each of my command, and output each step in a JSON object.
+
+const useWebVMProvider = ({
+  manualBoot = false,
+  workdirName = "ai-container",
+  welcomePrompt = `Welcome to AI Container: 
+  - Use "ai [prompt]" to issue any instructions
+  - Use "code [prompt]" to automate software engineering
+  - Non-standard unix commands will be simulated
+  `
+}) => {
+  const terminalRef = useRef<Terminal>()
+  const fitAddonRef = useRef<FitAddon>()
+  const containerRef = useRef<WebContainer>()
+  const shellProcRef = useRef<WebContainerProcess>()
+  const bootingRef = useRef(false)
+  const inputRef = useRef<WritableStreamDefaultWriter<string>>()
+
+  const codeAI = useCodeAI()
+
+  const shellAI = useWindowAI(
+    [
+      {
+        role: "system",
+        content: `You are a powerful Linux quantum virtual machine that can simulate new command-line program, non-standard UNIX program, and anything above and beyond. If a program can be answered with just text, write just text. If a program is creating graphic, you will use ASCII art. If a program is creating audio, you will write music notations or lyrics. Be super creative and resourceful with how you come up with the output. You are not just any text-based program. Do not give any explanation, apologies, or any reasoning. Just output the result.`
```

Comment:
nit, grammar: "simulate creating a new command-line program..."

Reply:
Tweaked it a little bit - "create and output"